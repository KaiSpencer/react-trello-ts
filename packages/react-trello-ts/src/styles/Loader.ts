import styled, {keyframes} from 'styled-components'

const keyframeAnimation = keyframes`
    0% {
      transform: scale(1);
    }
    20% {
      transform: scale(1, 2.2);
    }
    40% {
      transform: scale(1);
    }
`
export const LoaderDiv = styled.div`
  text-align: center;
  margin: 15px 0;
`

export const LoadingBar = styled.div`
  display: inline-block;
  margin: 0 2px;
  width: 4px;
  height: 18px;
  border-radius: 4px;
  animation: ${keyframeAnimation} 1s ease-in-out infinite;
  background-color: #777;

  &:nth-child(1) {
    animation-delay: 0.0001s;
  }
  &:nth-child(2) {
    animation-delay: 0.09s;
  }
  &:nth-child(3) {
    animation-delay: 0.18s;
  }
  &:nth-child(4) {
    animation-delay: 0.27s;
  }
`
