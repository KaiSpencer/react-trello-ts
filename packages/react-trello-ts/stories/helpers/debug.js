export default (message) => {
  if (process.env.NODE_ENV === 'test') { return }
  if (typeof message === 'object') {
    console.dir(message)
  } else {
    console.log(message)
  }
}
