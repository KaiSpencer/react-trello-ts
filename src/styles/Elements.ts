import styled from 'styled-components'
import {CardWrapper, MovableCardWrapper} from './Base'

export const DeleteWrapper = styled.div`
  text-align: center;
  position: absolute;
  top: -1px;
  right: 2px;
  cursor: pointer;
`

export const GenDelButton = styled.button`
  transition: all 0.5s ease;
  display: inline-block;
  border: none;
  font-size: 15px;
  height: 15px;
  padding: 0;
  margin-top: 5px;
  text-align: center;
  width: 15px;
  background: inherit;
  cursor: pointer;
`

export const DelButton = styled.button`
  transition: all 0.5s ease;
  display: inline-block;
  border: none;
  font-size: 8px;
  height: 15px;
  line-height: 1px;
  margin: 0 0 8px;
  padding: 0;
  text-align: center;
  width: 15px;
  background: inherit;
  cursor: pointer;
  opacity: 0;
  ${MovableCardWrapper}:hover & {
    opacity: 1;
  }
`

export const MenuButton = styled.button`
  transition: all 0.5s ease;
  display: inline-block;
  border: none;
  outline: none;
  font-size: 16px;
  font-weight: bold;
  height: 15px;
  line-height: 1px;
  margin: 0 0 8px;
  padding: 0;
  text-align: center;
  width: 15px;
  background: inherit;
  cursor: pointer;
`

export const LaneMenuHeader = styled.div`
    position: relative;
    margin-bottom: 4px;
    text-align: center;
`

export const LaneMenuContent = styled.div`
    overflow-x: hidden;
    overflow-y: auto;
    padding: 0 12px 12px;
`

export const LaneMenuItem = styled.div`
    cursor: pointer;
    display: block;
    font-weight: 700;
    padding: 6px 12px;
    position: relative;
    margin: 0 -12px;
    text-decoration: none;

    &:hover {
      background-color: #3179BA;
      color: #fff;
    }
`

export const LaneMenuTitle = styled.span`
    box-sizing: border-box;
    color: #6b808c;
    display: block;
    line-height: 30px;
    border-bottom: 1px solid rgba(9,45,66,.13);
    margin: 0 6px;
    overflow: hidden;
    padding: 0 32px;
    position: relative;
    text-overflow: ellipsis;
    white-space: nowrap;
    z-index: 1;
`

export const DeleteIcon = styled.span`
  position: relative;
  display: inline-block;
  width: 4px;
  height: 4px;
  opacity: 1;
  overflow: hidden;
  border: 1px solid #83bd42;
  border-radius: 50%;
  padding: 4px;
  background-color: #83bd42;

  ${CardWrapper}:hover & {
    opacity: 1;
  }

  &:hover::before,
  &:hover::after {
    background: red;
  }

  &:before,
  &:after {
    content: '';
    position: absolute;
    height: 2px;
    width: 60%;
    top: 45%;
    left: 20%;
    background: #fff;
    border-radius: 5px;
  }

  &:before {
    -webkit-transform: rotate(45deg);
    -moz-transform: rotate(45deg);
    -o-transform: rotate(45deg);
    transform: rotate(45deg);
  }

  &:after {
    -webkit-transform: rotate(-45deg);
    -moz-transform: rotate(-45deg);
    -o-transform: rotate(-45deg);
    transform: rotate(-45deg);
  }
`

export const ExpandCollapseBase = styled.span`
  width: 36px;
  margin: 0 auto;
  font-size: 14px;
  position: relative;
  cursor: pointer;
`

export const CollapseBtn = styled(ExpandCollapseBase)`
  &:before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    border-bottom: 7px solid #444;
    border-left: 7px solid transparent;
    border-right: 7px solid transparent;
    border-radius: 6px;
  }
  &:after {
    content: '';
    position: absolute;
    left: 4px;
    top: 4px;
    border-bottom: 3px solid #e3e3e3;
    border-left: 3px solid transparent;
    border-right: 3px solid transparent;
  }
`

export const ExpandBtn = styled(ExpandCollapseBase)`
  &:before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    border-top: 7px solid #444;
    border-left: 7px solid transparent;
    border-right: 7px solid transparent;
    border-radius: 6px;
  }
  &:after {
    content: '';
    position: absolute;
    left: 4px;
    top: 0px;
    border-top: 3px solid #e3e3e3;
    border-left: 3px solid transparent;
    border-right: 3px solid transparent;
  }
`

export const AddButton = styled.button`
  background: #5aac44;
  color: #fff;
  transition: background 0.3s ease;
  min-height: 32px;
  padding: 4px 16px;
  vertical-align: top;
  margin-top: 0;
  margin-right: 8px;
  font-weight: bold;
  border-radius: 3px;
  font-size: 14px;
  cursor: pointer;
  margin-bottom: 0;
`

export const CancelButton = styled.button`
  background: #999999;
  color: #fff;
  transition: background 0.3s ease;
  min-height: 32px;
  padding: 4px 16px;
  vertical-align: top;
  margin-top: 0;
  font-weight: bold;
  border-radius: 3px;
  font-size: 14px;
  cursor: pointer;
  margin-bottom: 0;
`
export const AddLaneLink = styled.button`
  background: #2b6aa3;
  border: none;
  color: #fff;
  transition: background 0.3s ease;
  min-height: 32px;
  padding: 4px 16px;
  vertical-align: top;
  margin-top: 0;
  margin-right: 0px;
  border-radius: 4px;
  font-size: 13px;
  cursor: pointer;
  margin-bottom: 0;
`
