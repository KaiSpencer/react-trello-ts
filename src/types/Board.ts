import {CSSProperties} from 'react'

export interface BoardData {
  lanes: Lane[]
}

export interface Lane {
  id: string
  title?: string
  label?: string
  style?: CSSProperties
  cards: Card[]
  currentPage?: number
  droppable?: boolean
  labelStyle?: CSSProperties
  cardStyle?: CSSProperties
  disallowAddingCard?: boolean
  [key: string]: any
}

export interface Card {
  id: string
  title?: string
  label?: string
  description?: string
  laneId?: string
  style?: CSSProperties
  draggable?: boolean
  [key: string]: any
}
