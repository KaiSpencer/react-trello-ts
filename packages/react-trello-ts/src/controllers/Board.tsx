import React, { createContext, FC, PropsWithChildren } from "react";
import classNames from "classnames";
import uuidv1 from "uuid/v1";
import { BoardContainer } from "./BoardContainer";
import * as DefaultComponets from "../components";
import { BoardData } from "../types/Board";
import { store } from "../store/store";

export const Board: FC<
	PropsWithChildren<{
		id?: string;
		className?: string;
		components: typeof DefaultComponets;
		data: BoardData;
		t?: any;
	}>
> = ({ data, children, className, components, id, t, ...rest }) => {
	const allClassNames = classNames("react-trello-board", className || "");

	return (
		<BoardContext.Provider value={store}>
			<components.GlobalStyle />
			<BoardContainer
				id={id || uuidv1()}
				data={data}
				components={components}
				t={t}
				className={allClassNames}
				{...rest}
			/>
		</BoardContext.Provider>
	);
};

export const BoardContext = createContext(store);
