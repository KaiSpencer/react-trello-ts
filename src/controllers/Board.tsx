import React, {createContext, FC, PropsWithChildren} from 'react'
import classNames from 'classnames'
import uuidv1 from 'uuid/v1'
import {BoardContainer} from './BoardContainer'
import * as DefaultComponets from '../components'
import {BoardData} from 'rt/types/Board'
import {store} from 'rt/store/store'

// export class _Board extends Component<{
//   id?: string
//   className?: string
//   components?: typeof DefaultComponets
//   data: BoardData
//   t?: any
// }> {
//   store: any
//   id: any
//   constructor(props) {
//     super(props)
//     this.store = this.getStore()
//     this.id = this.props.id || uuidv1()
//   }

//   getStore = () => {
//     //When you create multiple boards, unique stores are created for isolation
//     return createStore(boardReducer)
//   }

//   render() {
//     const {className, components} = this.props
//     const allClassNames = classNames('react-trello-board', className || '')
//     return (
//       <Provider store={this.store}>
//         <>
//           <components.GlobalStyle />
//           <BoardContainer id={this.id} {...this.props} className={allClassNames} />
//         </>
//       </Provider>
//     )
//   }
// }

export const Board: FC<
  PropsWithChildren<{
    id?: string
    className?: string
    components: typeof DefaultComponets
    data: BoardData
    t?: any
  }>
> = ({data, children, className, components, id, t, ...rest}) => {
  const allClassNames = classNames('react-trello-board', className || '')

  return (
    <BoardContext.Provider value={store}>
      <components.GlobalStyle />
      <BoardContainer
        id={id || uuidv1()}
        data={data}
        components={components}
        t={t}
        className={allClassNames}
        {...rest}
      />
    </BoardContext.Provider>
  )
}

export const BoardContext = createContext(store)
