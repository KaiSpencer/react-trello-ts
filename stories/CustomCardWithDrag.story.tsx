import React, {Component} from 'react'
import {storiesOf} from '@storybook/react'
import update from 'immutability-helper'

import {MovableCardWrapper} from 'rt/styles/Base'
import debug from './helpers/debug'

import Board from '../src'

const CustomCard = props => {
  return (
    <MovableCardWrapper
      data-id={props.id}
      onClick={props.onClick}
      style={props.style}
      className={props.className}
      style={{backgroundColor: props.cardColor, padding: 6}}>
      <header
        style={{
          borderBottom: '1px solid #eee',
          paddingBottom: 6,
          marginBottom: 10,
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between'
        }}>
        <div style={{fontSize: 14, fontWeight: 'bold'}}>{props.name}</div>
      </header>
      <div style={{fontSize: 12, color: '#BD3B36'}}>
        <div style={{color: '#4C4C4C', fontWeight: 'bold'}}>{props.subTitle}</div>
        <div style={{padding: '5px 0px'}}>
          <i>{props.body}</i>
        </div>
      </div>
    </MovableCardWrapper>
  )
}

const customCardData = {
  lanes: [
    {
      id: 'lane1',
      title: 'Planned',
      cards: [
        {
          id: 'Card1',
          name: 'John Smith',
          subTitle: 'SMS received at 12:13pm today',
          body: 'Thanks. Please schedule me for an estimate on Monday.',
          metadata: {id: 'Card1'}
        },
        {
          id: 'Card2',
          name: 'Card Weathers',
          subTitle: 'Email received at 1:14pm',
          body: 'Is the estimate free, and can someone call me soon?',
          metadata: {id: 'Card1'}
        }
      ]
    },
    {
      id: 'lane2',
      title: 'Work In Progress',
      cards: [
        {
          id: 'Card3',
          name: 'Michael Caine',
          subTitle: 'Email received at 4:23pm today',
          body: 'You are welcome. Interested in doing business with you again',
          metadata: {id: 'Card1'}
        }
      ]
    }
  ]
}

class BoardWithCustomCard extends Component {
  state = {boardData: customCardData, draggedData: undefined}

  updateBoard = newData => {
    debug('calling updateBoard')
    this.setState({draggedData: newData})
  }

  onDragEnd = (cardId, sourceLandId, targetLaneId, card) => {
    debug('Calling onDragENd')
    const {draggedData} = this.state
    const laneIndex = draggedData.lanes.findIndex(lane => lane.id === sourceLandId)
    const cardIndex = draggedData.lanes[laneIndex].cards.findIndex(card => card.id === cardId)
    const updatedData = update(draggedData, {
      lanes: {[laneIndex]: {cards: {[cardIndex]: {cardColor: {$set: '#d0fdd2'}}}}}
    })
    this.setState({boardData: updatedData})
  }

  render() {
    return (
      <Board
        tagStyle={{fontSize: '80%'}}
        data={this.state.boardData}
        draggable
        onDataChange={this.updateBoard}
        handleDragEnd={this.onDragEnd}
        onCardClick={(cardId, metadata) => alert(`Card with id:${cardId} clicked. Has metadata.id: ${metadata.id}`)}
        components={{Card: CustomCard}}
      />
    )
  }
}

storiesOf('Custom Components', module).add(
  'Drag-n-Drop Styling',
  () => {
    return <BoardWithCustomCard />
  },
  {info: 'Change card color on drag-n-drop'}
)
