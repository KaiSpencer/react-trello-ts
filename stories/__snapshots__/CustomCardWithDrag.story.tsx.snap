// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Components DragnDropStyling smoke-test 1`] = `
<div draggable="false"
     class="sc-iGgWBj iSNfA"
>
  <div>
    <div class="smooth-dnd-container horizontal">
      <div class="smooth-dnd-draggable-wrapper">
        <section draggable="false"
                 class="sc-kAyceB dgWEjS react-trello-lane"
        >
          <header class="sc-gsFSXq sc-imWYAI imRIDz gLxuAH">
            <span class="sc-kpDqfm gDkBAH">
              Planned
            </span>
          </header>
          <div class="sc-dhKdcB ikKiDc">
            <div class="smooth-dnd-container vertical">
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="Card1"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                         style="padding: 6px;"
                >
                  <header style="border-bottom: 1px solid rgb(238, 238, 238); padding-bottom: 6px; margin-bottom: 10px; display: flex; flex-direction: row; justify-content: space-between;">
                    <div style="font-size: 14px; font-weight: bold;">
                      John Smith
                    </div>
                  </header>
                  <div style="font-size: 12px; color: rgb(189, 59, 54);">
                    <div style="color: rgb(76, 76, 76); font-weight: bold;">
                      SMS received at 12:13pm today
                    </div>
                    <div style="padding: 5px 0px;">
                      <i>
                        Thanks. Please schedule me for an estimate on Monday.
                      </i>
                    </div>
                  </div>
                </article>
              </div>
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="Card2"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                         style="padding: 6px;"
                >
                  <header style="border-bottom: 1px solid rgb(238, 238, 238); padding-bottom: 6px; margin-bottom: 10px; display: flex; flex-direction: row; justify-content: space-between;">
                    <div style="font-size: 14px; font-weight: bold;">
                      Card Weathers
                    </div>
                  </header>
                  <div style="font-size: 12px; color: rgb(189, 59, 54);">
                    <div style="color: rgb(76, 76, 76); font-weight: bold;">
                      Email received at 1:14pm
                    </div>
                    <div style="padding: 5px 0px;">
                      <i>
                        Is the estimate free, and can someone call me soon?
                      </i>
                    </div>
                  </div>
                </article>
              </div>
            </div>
          </div>
        </section>
      </div>
      <div class="smooth-dnd-draggable-wrapper">
        <section draggable="false"
                 class="sc-kAyceB dgWEjS react-trello-lane"
        >
          <header class="sc-gsFSXq sc-imWYAI imRIDz gLxuAH">
            <span class="sc-kpDqfm gDkBAH">
              Work In Progress
            </span>
          </header>
          <div class="sc-dhKdcB ikKiDc">
            <div class="smooth-dnd-container vertical">
              <div class="smooth-dnd-draggable-wrapper">
                <article data-id="Card3"
                         class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
                         style="padding: 6px;"
                >
                  <header style="border-bottom: 1px solid rgb(238, 238, 238); padding-bottom: 6px; margin-bottom: 10px; display: flex; flex-direction: row; justify-content: space-between;">
                    <div style="font-size: 14px; font-weight: bold;">
                      Michael Caine
                    </div>
                  </header>
                  <div style="font-size: 12px; color: rgb(189, 59, 54);">
                    <div style="color: rgb(76, 76, 76); font-weight: bold;">
                      Email received at 4:23pm today
                    </div>
                    <div style="padding: 5px 0px;">
                      <i>
                        You are welcome. Interested in doing business with you again
                      </i>
                    </div>
                  </div>
                </article>
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </div>
</div>
`;
